// @whiskeysockets/baileys/lib/Types/Newsletter.js
export const MexOperations = {
    PROMOTE: 'NotificationNewsletterAdminPromote',
    DEMOTE: 'NotificationNewsletterAdminDemote',
    UPDATE: 'NotificationNewsletterUpdate'
};

export var XWAPaths;
(function (XWAPaths) {
    XWAPaths["xwa2_newsletter_create"] = "xwa2_newsletter_create";
    XWAPaths["xwa2_newsletter_subscribers"] = "xwa2_newsletter_subscribers";
    XWAPaths["xwa2_newsletter_subscribed"] = "xwa2_newsletter_subscribed";
    XWAPaths["xwa2_newsletter_view"] = "xwa2_newsletter_view";
    XWAPaths["xwa2_newsletter_metadata"] = "xwa2_newsletter";
    XWAPaths["xwa2_newsletter_admin_count"] = "xwa2_newsletter_admin";
    XWAPaths["xwa2_newsletter_mute_v2"] = "xwa2_newsletter_mute_v2";
    XWAPaths["xwa2_newsletter_unmute_v2"] = "xwa2_newsletter_unmute_v2";
    XWAPaths["xwa2_newsletter_follow"] = "xwa2_newsletter_follow";
    XWAPaths["xwa2_newsletter_unfollow"] = "xwa2_newsletter_unfollow";
    XWAPaths["xwa2_newsletter_change_owner"] = "xwa2_newsletter_change_owner";
    XWAPaths["xwa2_newsletter_demote"] = "xwa2_newsletter_demote";
    XWAPaths["xwa2_newsletter_delete_v2"] = "xwa2_newsletter_delete_v2";
    XWAPaths["PROMOTE"] = "xwa2_notify_newsletter_admin_promote";
    XWAPaths["DEMOTE"] = "xwa2_notify_newsletter_admin_demote";
    XWAPaths["NEWSLETTER"] = "xwa2_newsletter";
    XWAPaths["SUBSCRIBED"] = "xwa2_newsletter_subscribed";
    XWAPaths["METADATA_UPDATE"] = "xwa2_notify_newsletter_on_metadata_update";
    XWAPaths["CREATE"] = "xwa2_newsletter_create";
    XWAPaths["ADMIN_COUNT"] = "xwa2_newsletter_admin";
})(XWAPaths || (XWAPaths = {}));

export var QueryIds;
(function (QueryIds) {
    QueryIds["JOB_MUTATION"] = "7150902998257522";
    QueryIds["CREATE"] = "6996806640408138";
    QueryIds["METADATA"] = "6620195908089573";
    QueryIds["SUBSCRIBERS"] = "7977329212281851";
    QueryIds["FOLLOW"] = "7871414976211147";
    QueryIds["UNFOLLOW"] = "7238632346214362";
    QueryIds["MUTE"] = "25151904754424642";
    QueryIds["UNMUTE"] = "7337137176362961";
    QueryIds["ADMIN_COUNT"] = "7130823597031706";
    QueryIds["CHANGE_OWNER"] = "7341777602580933";
    QueryIds["DEMOTE"] = "6551828931592903";
    QueryIds["DELETE"] = "8316537688363079";
    QueryIds["SUBSCRIBED"] = "9926858900865316";
    QueryIds["UPDATE_METADATA"] = "7150902998257522";
})(QueryIds || (QueryIds = {}));
